File: packages/ycaptool/cli/capture.py
'''
    Copyright (C) 2025  a7mddra-spatialshot
  
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
  
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
  
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
'''

import os
import sys
import shutil
import logging
from pathlib import Path
from typing import Tuple

from utils import run_cmd
from audio import AudioManager
from notify import NotificationManager

logger = logging.getLogger("ycaptool.cli.capture")


class Capture:
    def __init__(self):
        self.audio_manager = AudioManager()
        self.notify_manager = NotificationManager()

    def get_flameshot_path(self) -> str:
        try:
            if getattr(sys, "frozen", False):
                bundle_dir = Path(sys._MEIPASS)
            else:
                bundle_dir = Path(__file__).parent

            bundled = bundle_dir / "flameshot"
            if bundled.exists():
                try:
                    bundled.chmod(0o755)
                except Exception:
                    pass
                return str(bundled)

            system_path = shutil.which("flameshot")
            if system_path:
                logger.warning("Using system-installed flameshot: %s", system_path)
                return system_path

            logger.error("No flameshot executable found. Expected %s", bundled)
            raise SystemExit(1)
        except Exception as exc:
            logger.error("Error locating flameshot: %s", exc)
            raise SystemExit(1)

    def capture_display(self, display_num: int) -> Tuple[bool, str]:
        if display_num <= 0:
            return False, "invalid display number"

        SC_SAVE_PATH = Path.home() / ".config" / "spatialshot" / "tmp"
        save_path = Path(os.environ.get("SC_SAVE_PATH", SC_SAVE_PATH))
        save_path.mkdir(parents=True, exist_ok=True)
        image_fmt = "png"
        output_file = save_path / f"{display_num}.{image_fmt}"

        flameshot_path = self.get_flameshot_path()
        flameshot_index = display_num - 1

        try:
            self.audio_manager.mute_audio()
            self.notify_manager.enable_dnd()
        except Exception as exc:
            logger.warning("Could not mute audio: %s", exc)

        try:
            rc, out = run_cmd([
                                flameshot_path, "screen", "-n",
                                str(flameshot_index), "--path",
                                str(output_file)
                                ], capture=True)
            if rc == 0:
                return True, f"Saved: {output_file}"
            msg = f"Capture failed for display {display_num} (rc={rc})"
            if out:
                msg += f": {out}"
            return False, msg
        except Exception as exc:
            return False, f"Exception during capture: {exc}"
        finally:
            try:
                self.audio_manager.restore_audio()
                self.notify_manager.restore_dnd()
            except Exception as exc:
                logger.warning("Could not restore audio: %s", exc)

File: packages/ycaptool/src/capture.cpp
/**
 *  Copyright (C) 2025  a7mddra-spatialshot
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
**/

#include "capture.h"
#include "audio_manager.h"
#include "dnd_manager.h"
#include "shell.h"
#include <iostream>
#include <string>
#include <vector>
#include <cstdlib>
#include <unistd.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <cstring>
#include <gio/gio.h>

static std::string extract_resource_to_temp(const char *resource_path)
{
    GError *error = nullptr;
    GBytes *bytes = g_resources_lookup_data(resource_path, G_RESOURCE_LOOKUP_FLAGS_NONE, &error);
    if (!bytes)
    {
        if (error) {
            std::cerr << "Resource lookup failed for " << resource_path << ": " << error->message << std::endl;
            g_error_free(error);
        }
        return "";
    }

    gsize size = 0;
    const void *data = g_bytes_get_data(bytes, &size);

    char temp_template[] = "/tmp/ycap-exec-XXXXXX";
    int fd = mkstemp(temp_template);
    if (fd == -1)
    {
        std::cerr << "mkstemp failed: " << strerror(errno) << std::endl;
        g_bytes_unref(bytes);
        return "";
    }

    ssize_t written = write(fd, data, size);
    close(fd);
    g_bytes_unref(bytes);

    if (written != (ssize_t)size)
    {
        std::cerr << "Failed to write resource to temp file" << std::endl;
        unlink(temp_template);
        return "";
    }

    if (chmod(temp_template, 0755) != 0)
    {
        std::cerr << "chmod failed on temp file: " << strerror(errno) << std::endl;
        unlink(temp_template);
        return "";
    }

    return std::string(temp_template);
}
static std::string get_flameshot_path()
{
    std::string bundled_path = extract_resource_to_temp("/org/spatialshot/ycaptool/bin/flameshot");
    if (!bundled_path.empty())
    {
        return bundled_path;
    }

    if (Shell::command_exists("flameshot"))
    {
        std::string system_path = Shell::run_and_get_output("which flameshot");
        if (!system_path.empty()) {
            std::cerr << "Using system-installed flameshot: " << system_path << std::endl;
            return system_path;
        }
    }

    std::cerr << "Flameshot executable not found." << std::endl;
    return "";
}

void capture_screen(int display_number)
{
    if (display_number < 0)
    {
        std::cerr << "Invalid display number: " << display_number << std::endl;
        return;
    }

    std::string flameshot_path = get_flameshot_path();
    if (flameshot_path.empty())
    {
        return;
    }

    AudioManager audio_manager;
    DndManager dnd_manager;

    audio_manager.mute_audio();
    dnd_manager.enable_dnd();

    std::string cache_base_path;
    const char* xdg_cache_home = std::getenv("XDG_CACHE_HOME");

    if (xdg_cache_home && xdg_cache_home[0] != '\0') {
        cache_base_path = xdg_cache_home;
    } else {
        const char* home_dir = std::getenv("HOME");
        if (home_dir && home_dir[0] != '\0') {
            cache_base_path = std::string(home_dir) + "/.cache";
        } else {
            std::cerr << "Error: Cannot find $XDG_CACHE_HOME or $HOME. Unable to determine cache directory." << std::endl;
            audio_manager.restore_audio();
            dnd_manager.restore_dnd();
            return;
        }
    }

    std::string save_path_str = cache_base_path + "/spatialshot/tmp";
    Shell::run_silent("mkdir -p " + save_path_str);
    std::string output_file = save_path_str + "/" + std::to_string(display_number) + ".png";

    std::string cmd = flameshot_path + " screen -n " + std::to_string(display_number > 0 ? display_number - 1 : 0) + " --path " + output_file;

    int exit_code = Shell::run_silent(cmd);

    if (exit_code != 0)
    {
        std::cerr << "Flameshot capture failed with exit code: " << exit_code << std::endl;
    }
    else
    {
        std::cout << "Saved: " << output_file << std::endl;
    }

    audio_manager.restore_audio();
    dnd_manager.restore_dnd();
}
